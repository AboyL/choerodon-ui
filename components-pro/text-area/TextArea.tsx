import React, { ReactNode } from 'react';
import { observer } from 'mobx-react';
import isString from 'lodash/isString';
import omit from 'lodash/omit';
import PropTypes from 'prop-types';
import { TextField, TextFieldProps } from '../text-field/TextField';
import { ResizeType, AutoSizeType } from './enum';
import calculateNodeHeight from './calculateNodeHeight';

export interface TextAreaProps extends TextFieldProps {
  cols?: number;
  rows?: number;
  resize?: ResizeType;
  autoSize?: boolean | AutoSizeType;
  placeholder?: string;
}

@observer
export default class TextArea<T extends TextAreaProps> extends TextField<T> {
  static displayName = 'TextArea';

  static propTypes = {
    cols: PropTypes.number,
    rows: PropTypes.number,
    resize: PropTypes.oneOf([
      ResizeType.vertical,
      ResizeType.horizontal,
      ResizeType.none,
      ResizeType.both,
    ]),
    autoSize: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
    ...TextField.propTypes,
  };

  static defaultProps = {
    ...TextField.defaultProps,
    suffixCls: 'textarea',
    resize: ResizeType.none,
    rows: 4,
    autoSize: false,
    placeholder: '',
  };

  textAreaRef: HTMLTextAreaElement;

  saveTextArea = (textArea: HTMLTextAreaElement) => {
    this.textAreaRef = textArea;
  };

  getOtherProps() {
    const { resize } = this.props;
    const otherProps = omit(super.getOtherProps(), ['resize', 'autoSize']);
    const { style = {} } = otherProps;
    style.resize = resize;
    if (resize !== ResizeType.none) {
      style.transition = 'none';
    }
    const { autoSize } = this.props;
    let textAreaStyles = {};
    if (autoSize) {
      const { minRows, maxRows } = autoSize as AutoSizeType;
      otherProps.rows = minRows;
      if (this.textAreaRef) {
        textAreaStyles = calculateNodeHeight(this.textAreaRef, true, minRows, maxRows);
      }
    }
    otherProps.style = { ...style, ...textAreaStyles };
    return otherProps;
  }

  renderWrapper(): ReactNode {
    const text = this.getTextNode();
    return (
      <div key="wrapper" {...this.getWrapperProps()}>
        {this.renderPlaceHolder()}
        <label>
          <textarea
            {...this.getOtherProps()}
            placeholder={this.hasFloatLabel ? undefined : this.getPlaceholders()[0]}
            readOnly={!this.editable}
            ref={this.saveTextArea}
            value={isString(text) ? text : this.getText(this.getValue())}
          />
          {this.renderFloatLabel()}
        </label>
      </div>
    );
  }

  handleEnterDown() {}
}
