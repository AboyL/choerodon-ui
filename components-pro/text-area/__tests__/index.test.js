import React from 'react';
import { mount } from 'enzyme';
import TextArea from '..';
import TextAreaTest from './textAreaTest';

describe('TextArea-pro', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('renders TextArea-pro correctly', () => {
    const wrapper = mount(<TextArea />);
    expect(wrapper).toMatchSnapshot();
  });

  it('TextArea should have the default resize and rows', () => {
    const wrapper = mount(<TextArea />);
    expect(wrapper.prop('resize')).toBe('none');
    expect(wrapper.prop('rows')).toBe(4);
  });

  it('TextArea have the disabled property and default false', () => {
    const wrapper = mount(<TextArea />);
    expect(wrapper.find('.c7n-pro-textarea-disabled')).toHaveLength(0);
    wrapper.setProps({ disabled: true });
    wrapper.update();
    expect(wrapper.find('.c7n-pro-textarea-disabled')).toHaveLength(1);
  });

  it('can not click or input when the readOnly true and default false', () => {
    const handleChange = jest.fn();
    const wrapper = mount(<TextArea onChange={handleChange} />);
    expect(wrapper.prop('readOnly')).toBe(false);
    wrapper.setProps({ readOnly: true });
    wrapper.update();
    expect(wrapper.prop('readOnly')).toBe(true);

    wrapper.find('textarea').simulate('change');
    jest.runAllTimers();
    wrapper.update();
    expect(handleChange).not.toHaveBeenCalled();
  });

  it('the value will be controlled and the change event trigger when the value change', () => {
    const handleChange = jest.fn();
    const wrapper = mount(<TextArea value="text" onChange={handleChange} />);
    expect(
      wrapper
        .find('textarea')
        .at(0)
        .prop('value'),
    ).toBe('text');
    wrapper.setProps({ value: 'text-area' });
    wrapper.update();
    expect(
      wrapper
        .find('textarea')
        .at(0)
        .prop('value'),
    ).toBe('text-area');
  });

  it('the dataSet should renders correctly', () => {
    const wrapper = mount(<TextAreaTest />);
    expect(wrapper.find('TextArea').props().name).toEqual('content');
    expect(wrapper).toMatchSnapshot();
  });
});
