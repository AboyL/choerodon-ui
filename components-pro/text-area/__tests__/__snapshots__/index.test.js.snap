// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea-pro renders TextArea-pro correctly 1`] = `
<TextArea
  autoComplete="off"
  clearButton={false}
  multiple={false}
  noValidate={false}
  readOnly={false}
  resize="none"
  rows={4}
  showHelp="newLine"
  suffixCls="textarea"
  trim="both"
>
  <Tooltip
    arrowPointAtCenter={false}
    autoAdjustOverflow={true}
    defaultHidden={true}
    mouseEnterDelay={100}
    mouseLeaveDelay={100}
    placement="bottomLeft"
    suffixCls="tooltip"
    theme="light"
    title={null}
    transitionName="zoom-big-fast"
    trigger={
      Array [
        "hover",
      ]
    }
  >
    <Trigger
      action={
        Array [
          "hover",
        ]
      }
      arrowPointAtCenter={false}
      autoAdjustOverflow={true}
      blurDelay={0}
      builtinPlacements={
        Object {
          "bottom": Object {
            "offset": Array [
              0,
              10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tc",
              "bc",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "bottomLeft": Object {
            "offset": Array [
              0,
              10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tl",
              "bl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "bottomRight": Object {
            "offset": Array [
              0,
              10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tr",
              "br",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "left": Object {
            "offset": Array [
              -10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "cr",
              "cl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "leftBottom": Object {
            "offset": Array [
              -10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "br",
              "bl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "leftTop": Object {
            "offset": Array [
              -10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tr",
              "tl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "right": Object {
            "offset": Array [
              10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "cl",
              "cr",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "rightBottom": Object {
            "offset": Array [
              10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "bl",
              "br",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "rightTop": Object {
            "offset": Array [
              10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tl",
              "tr",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "top": Object {
            "offset": Array [
              0,
              -10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "bc",
              "tc",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "topLeft": Object {
            "offset": Array [
              0,
              -10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "bl",
              "tl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "topRight": Object {
            "offset": Array [
              0,
              -10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "br",
              "tr",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
        }
      }
      defaultPopupHidden={true}
      focusDelay={150}
      mouseEnterDelay={100}
      mouseLeaveDelay={100}
      popupContent={null}
      popupPlacement="bottomLeft"
      prefixCls="c7n-pro-tooltip"
      suffixCls="tooltip"
      theme="light"
      title={null}
      transitionName="zoom-big-fast"
    >
      <TriggerChild
        key=".$.$wrapper"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="c7n-pro-textarea-wrapper c7n-pro-textarea-empty"
          key=".$wrapper"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <label>
            <textarea
              autoComplete="off"
              className="c7n-pro-textarea"
              multiple={false}
              noValidate={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              readOnly={false}
              rows={4}
              style={
                Object {
                  "resize": "none",
                }
              }
              type="text"
              value=""
            />
          </label>
        </div>
      </TriggerChild>
      <Popup
        align={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tl",
              "bl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          }
        }
        className="c7n-pro-tooltip-popup"
        getClassNameFromAlign={[Function]}
        getRootDomNode={[Function]}
        hidden={true}
        key="popup"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        suffixCls="popup"
        transitionName="zoom-big-fast"
      />
    </Trigger>
  </Tooltip>
</TextArea>
`;
